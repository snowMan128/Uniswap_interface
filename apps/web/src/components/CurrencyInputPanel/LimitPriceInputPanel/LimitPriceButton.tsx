import { Trans } from '@lingui/macro'
import { Percent } from '@uniswap/sdk-core'
import Row from 'components/Row'
import { X } from 'react-feather'
import styled, { css } from 'styled-components'
import { ClickableStyle } from 'theme/components'
import { useFormatter } from 'utils/formatNumbers'

interface LimitPriceButtonProps {
  priceAdjustmentPercentage: number
  disabled?: boolean
  selected?: boolean
  onSelect: (priceAdjustmentPercentage: number) => void
}

const containerBorderCss = css`
  border-radius: 999px;
  border: 1px solid ${({ theme }) => theme.surface3};
`

const highlightedBorderCss = css`
  border-radius: 999px 0px 0px 999px;
  border-top: 1px solid ${({ theme }) => theme.surface3};
  border-bottom: 1px solid ${({ theme }) => theme.surface3};
  border-left: 1px solid ${({ theme }) => theme.surface3};
  border-right: 1px solid transparent;
`

const Container = styled.button<{ $selected?: boolean; $disabled?: boolean; $highlighted: boolean }>`
  background-color: ${({ theme, $highlighted }) => ($highlighted ? theme.surface3 : 'unset')};
  color: ${({ theme, $selected, $disabled, $highlighted }) =>
    $highlighted || ($selected && !$disabled) ? theme.neutral1 : theme.neutral2};
  padding: 2px 8px;
  height: 24px;
  display: flex;
  justify-content: center;
  align-items: center;
  background-color: ${({ theme, $selected, $highlighted }) =>
    $selected || $highlighted ? theme.surface3 : 'transparent'};
  ${({ $highlighted }) => ($highlighted ? highlightedBorderCss : containerBorderCss)};
  ${({ $disabled }) => !$disabled && ClickableStyle};
`

const HighlightedContainerXButton = styled.button`
  ${ClickableStyle}
  background-color: ${({ theme }) => theme.surface3};
  color: ${({ theme }) => theme.neutral2};
  display: flex;
  align-items: center;
  justify-content: center;
  text-align: center;
  padding: 6px 6px 6px 4px;
  height: 24px;
  border-radius: 0px 999px 999px 0px;
  border-top: 1px solid ${({ theme }) => theme.surface3};
  border-bottom: 1px solid ${({ theme }) => theme.surface3};
  border-right: 1px solid ${({ theme }) => theme.surface3};
  border-left: 1px solid transparent;
`

export function LimitPresetPriceButton({
  priceAdjustmentPercentage,
  selected,
  disabled,
  onSelect,
}: LimitPriceButtonProps) {
  const { formatPercent } = useFormatter()
  return (
    <Container
      $selected={selected}
      $disabled={disabled}
      $highlighted={false}
      onClick={() => !disabled && onSelect(priceAdjustmentPercentage)}
    >
      {priceAdjustmentPercentage === 0 ? (
        <Trans>Current</Trans>
      ) : (
        <span>+{formatPercent(new Percent(priceAdjustmentPercentage, 100))}</span>
      )}
    </Container>
  )
}

/**
 * A button to reset the price to the market price (i.e. an adjustment of 0%)
 * When defined, this button displays customAdjustmentPercentage instead of "Current"
 */
export function LimitCustomMarketPriceButton({
  customAdjustmentPercentage,
  selected,
  disabled,
  onSelect,
}: Omit<LimitPriceButtonProps, 'priceAdjustmentPercentage'> & { customAdjustmentPercentage?: number }) {
  const onSetAdjustmentPercentage = () => !disabled && onSelect(0)
  const { formatPercent } = useFormatter()

  return (
    <Row width="unset" gap="1px">
      <Container
        $selected={selected}
        $disabled={disabled}
        $highlighted={customAdjustmentPercentage !== undefined}
        onClick={onSetAdjustmentPercentage}
      >
        {!customAdjustmentPercentage ? (
          <Trans>Current</Trans>
        ) : (
          <span>
            {customAdjustmentPercentage > 0 ? '+' : ''}
            {formatPercent(new Percent(customAdjustmentPercentage, 100))}
          </span>
        )}
      </Container>
      {customAdjustmentPercentage && (
        <HighlightedContainerXButton onClick={onSetAdjustmentPercentage}>
          <X size={12} />
        </HighlightedContainerXButton>
      )}
    </Row>
  )
}
