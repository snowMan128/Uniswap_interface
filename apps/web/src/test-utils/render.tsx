import { MockedProvider } from '@apollo/client/testing'
import { QueryClient, QueryClientProvider } from '@tanstack/react-query'
import { queries } from '@testing-library/dom'
import { RenderOptions, render } from '@testing-library/react'
import { RenderHookOptions, WrapperComponent, renderHook } from '@testing-library/react-hooks'
import Web3Provider from 'components/Web3Provider'
import { AssetActivityProvider } from 'graphql/data/apollo/AssetActivityProvider'
import { TokenBalancesProvider } from 'graphql/data/apollo/TokenBalancesProvider'
import { BlockNumberContext } from 'lib/hooks/useBlockNumber'
import { PropsWithChildren, ReactElement, ReactNode } from 'react'
import { HelmetProvider } from 'react-helmet-async/lib/index'
import { Provider } from 'react-redux'
import { BrowserRouter } from 'react-router-dom'
import store from 'state'
import { ThemeProvider } from 'theme'
import { TamaguiProvider } from 'theme/tamaguiProvider'
import { UnitagUpdaterContextProvider } from 'uniswap/src/features/unitags/context'

const queryClient = new QueryClient()

const BLOCK_NUMBER_CONTEXT = { fastForward: () => {}, block: 1234, mainnetBlock: 1234 }
function MockedBlockNumberProvider({ children }: PropsWithChildren) {
  return <BlockNumberContext.Provider value={BLOCK_NUMBER_CONTEXT}>{children}</BlockNumberContext.Provider>
}

const WithProviders = ({ children }: { children?: ReactNode }) => {
  return (
    <HelmetProvider>
      <Provider store={store}>
        <QueryClientProvider client={queryClient}>
          <BrowserRouter>
            <Web3Provider>
              <MockedProvider showWarnings={false}>
                <AssetActivityProvider>
                  <TokenBalancesProvider>
                    <MockedBlockNumberProvider>
                      <UnitagUpdaterContextProvider>
                        <ThemeProvider>
                          <TamaguiProvider>{children}</TamaguiProvider>
                        </ThemeProvider>
                      </UnitagUpdaterContextProvider>
                    </MockedBlockNumberProvider>
                  </TokenBalancesProvider>
                </AssetActivityProvider>
              </MockedProvider>
            </Web3Provider>
          </BrowserRouter>
        </QueryClientProvider>
      </Provider>
    </HelmetProvider>
  )
}

type CustomRenderOptions = Omit<RenderOptions, 'wrapper'>
const customRender = (ui: ReactElement, options?: CustomRenderOptions) => {
  return render<typeof queries>(ui, { ...options, wrapper: WithProviders })
}

type CustomRenderHookOptions<Props> = Omit<RenderHookOptions<Props>, 'wrapper'>
const customRenderHook = <Result, Props>(
  hook: (initialProps: Props) => Result,
  options?: CustomRenderHookOptions<Props>
) => {
  return renderHook(hook, { ...options, wrapper: WithProviders as WrapperComponent<Props> })
}

// Testing utils may export *.
// eslint-disable-next-line no-restricted-syntax
export * from '@testing-library/react'
export { customRender as render, customRenderHook as renderHook }
