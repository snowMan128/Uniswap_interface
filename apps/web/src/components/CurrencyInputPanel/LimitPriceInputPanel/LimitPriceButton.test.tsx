import userEvent from '@testing-library/user-event'
import { act, render, screen } from 'test-utils/render'

import { LimitCustomMarketPriceButton, LimitPresetPriceButton } from './LimitPriceButton'

describe('LimitPresetPriceButton', () => {
  it.each([
    [0, false, false],
    [0, true, false],
    [0, false, true],
    [5, false, true],
    [5, true, false],
    [10, false, true],
  ])('renders %p correctly, disabled %p and selected %p', async (num, disabled, selected) => {
    const onSelect = jest.fn()
    const { container } = render(
      <LimitPresetPriceButton
        priceAdjustmentPercentage={num}
        onSelect={onSelect}
        disabled={disabled}
        selected={selected}
      />
    )
    await act(() => userEvent.click(screen.getByText(num > 0 ? `+${num}%` : 'Current')))
    expect(container.firstChild).toMatchSnapshot()
    expect(onSelect).toHaveBeenCalledTimes(disabled ? 0 : 1)
  })
})

describe('LimitCustomMarketPriceButton', () => {
  it('renders the custom amount correctly', async () => {
    const onSelect = jest.fn()
    const { container } = render(<LimitCustomMarketPriceButton customAdjustmentPercentage={10} onSelect={onSelect} />)
    await act(() => userEvent.click(screen.getByText('+10%')))
    expect(container.firstChild).toMatchSnapshot()
    expect(onSelect).toHaveBeenCalledWith(0)
  })

  it('renders the custom amount correctly, negative change', async () => {
    const onSelect = jest.fn()
    const { container } = render(<LimitCustomMarketPriceButton customAdjustmentPercentage={-10} onSelect={onSelect} />)
    await act(() => userEvent.click(screen.getByText('-10%')))
    expect(container.firstChild).toMatchSnapshot()
    expect(onSelect).toHaveBeenCalledWith(0)
  })

  it('renders disabled correctly', async () => {
    const onSelect = jest.fn()
    render(<LimitCustomMarketPriceButton customAdjustmentPercentage={10} onSelect={onSelect} disabled={true} />)
    await act(() => userEvent.click(screen.getByText('+10%')))
    expect(onSelect).toHaveBeenCalledTimes(0)
  })

  it('renders selected correctly', async () => {
    const onSelect = jest.fn()
    render(<LimitCustomMarketPriceButton customAdjustmentPercentage={10} onSelect={onSelect} selected={true} />)
    await act(() => userEvent.click(screen.getByText('+10%')))
    expect(onSelect).toHaveBeenCalledTimes(1)
  })
})
