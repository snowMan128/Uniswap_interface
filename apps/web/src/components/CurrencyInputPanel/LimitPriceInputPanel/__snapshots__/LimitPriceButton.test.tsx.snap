// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LimitCustomMarketPriceButton renders the custom amount correctly 1`] = `
.c0 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  width: unset;
}

.c1 {
  width: unset;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  padding: 0;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  gap: 1px;
}

.c2 {
  background-color: #22222212;
  color: #222222;
  padding: 2px 8px;
  height: 24px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background-color: #22222212;
  border-radius: 999px 0px 0px 999px;
  border-top: 1px solid #22222212;
  border-bottom: 1px solid #22222212;
  border-left: 1px solid #22222212;
  border-right: 1px solid transparent;
  -webkit-text-decoration: none;
  text-decoration: none;
  cursor: pointer;
  -webkit-transition-duration: 125ms;
  transition-duration: 125ms;
}

.c2:hover {
  opacity: 0.6;
}

.c2:active {
  opacity: 0.4;
}

.c3 {
  -webkit-text-decoration: none;
  text-decoration: none;
  cursor: pointer;
  -webkit-transition-duration: 125ms;
  transition-duration: 125ms;
  background-color: #22222212;
  color: #7D7D7D;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  text-align: center;
  padding: 6px 6px 6px 4px;
  height: 24px;
  border-radius: 0px 999px 999px 0px;
  border-top: 1px solid #22222212;
  border-bottom: 1px solid #22222212;
  border-right: 1px solid #22222212;
  border-left: 1px solid transparent;
}

.c3:hover {
  opacity: 0.6;
}

.c3:active {
  opacity: 0.4;
}

<div
  class="c0 c1"
  width="unset"
>
  <button
    class="c2"
  >
    <span>
      +
      10%
    </span>
  </button>
  <button
    class="c3"
  >
    <svg
      fill="none"
      height="12"
      stroke="currentColor"
      stroke-linecap="round"
      stroke-linejoin="round"
      stroke-width="2"
      viewBox="0 0 24 24"
      width="12"
      xmlns="http://www.w3.org/2000/svg"
    >
      <line
        x1="18"
        x2="6"
        y1="6"
        y2="18"
      />
      <line
        x1="6"
        x2="18"
        y1="6"
        y2="18"
      />
    </svg>
  </button>
</div>
`;

exports[`LimitCustomMarketPriceButton renders the custom amount correctly, negative change 1`] = `
.c0 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  width: unset;
}

.c1 {
  width: unset;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  padding: 0;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  gap: 1px;
}

.c2 {
  background-color: #22222212;
  color: #222222;
  padding: 2px 8px;
  height: 24px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background-color: #22222212;
  border-radius: 999px 0px 0px 999px;
  border-top: 1px solid #22222212;
  border-bottom: 1px solid #22222212;
  border-left: 1px solid #22222212;
  border-right: 1px solid transparent;
  -webkit-text-decoration: none;
  text-decoration: none;
  cursor: pointer;
  -webkit-transition-duration: 125ms;
  transition-duration: 125ms;
}

.c2:hover {
  opacity: 0.6;
}

.c2:active {
  opacity: 0.4;
}

.c3 {
  -webkit-text-decoration: none;
  text-decoration: none;
  cursor: pointer;
  -webkit-transition-duration: 125ms;
  transition-duration: 125ms;
  background-color: #22222212;
  color: #7D7D7D;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  text-align: center;
  padding: 6px 6px 6px 4px;
  height: 24px;
  border-radius: 0px 999px 999px 0px;
  border-top: 1px solid #22222212;
  border-bottom: 1px solid #22222212;
  border-right: 1px solid #22222212;
  border-left: 1px solid transparent;
}

.c3:hover {
  opacity: 0.6;
}

.c3:active {
  opacity: 0.4;
}

<div
  class="c0 c1"
  width="unset"
>
  <button
    class="c2"
  >
    <span>
      -10%
    </span>
  </button>
  <button
    class="c3"
  >
    <svg
      fill="none"
      height="12"
      stroke="currentColor"
      stroke-linecap="round"
      stroke-linejoin="round"
      stroke-width="2"
      viewBox="0 0 24 24"
      width="12"
      xmlns="http://www.w3.org/2000/svg"
    >
      <line
        x1="18"
        x2="6"
        y1="6"
        y2="18"
      />
      <line
        x1="6"
        x2="18"
        y1="6"
        y2="18"
      />
    </svg>
  </button>
</div>
`;

exports[`LimitPresetPriceButton renders 0 correctly, disabled false and selected false 1`] = `
.c0 {
  background-color: unset;
  color: #7D7D7D;
  padding: 2px 8px;
  height: 24px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background-color: transparent;
  border-radius: 999px;
  border: 1px solid #22222212;
  -webkit-text-decoration: none;
  text-decoration: none;
  cursor: pointer;
  -webkit-transition-duration: 125ms;
  transition-duration: 125ms;
}

.c0:hover {
  opacity: 0.6;
}

.c0:active {
  opacity: 0.4;
}

<button
  class="c0"
>
  Current
</button>
`;

exports[`LimitPresetPriceButton renders 0 correctly, disabled false and selected true 1`] = `
.c0 {
  background-color: unset;
  color: #222222;
  padding: 2px 8px;
  height: 24px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background-color: #22222212;
  border-radius: 999px;
  border: 1px solid #22222212;
  -webkit-text-decoration: none;
  text-decoration: none;
  cursor: pointer;
  -webkit-transition-duration: 125ms;
  transition-duration: 125ms;
}

.c0:hover {
  opacity: 0.6;
}

.c0:active {
  opacity: 0.4;
}

<button
  class="c0"
>
  Current
</button>
`;

exports[`LimitPresetPriceButton renders 0 correctly, disabled true and selected false 1`] = `
.c0 {
  background-color: unset;
  color: #7D7D7D;
  padding: 2px 8px;
  height: 24px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background-color: transparent;
  border-radius: 999px;
  border: 1px solid #22222212;
}

<button
  class="c0"
>
  Current
</button>
`;

exports[`LimitPresetPriceButton renders 5 correctly, disabled false and selected true 1`] = `
.c0 {
  background-color: unset;
  color: #222222;
  padding: 2px 8px;
  height: 24px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background-color: #22222212;
  border-radius: 999px;
  border: 1px solid #22222212;
  -webkit-text-decoration: none;
  text-decoration: none;
  cursor: pointer;
  -webkit-transition-duration: 125ms;
  transition-duration: 125ms;
}

.c0:hover {
  opacity: 0.6;
}

.c0:active {
  opacity: 0.4;
}

<button
  class="c0"
>
  <span>
    +
    5%
  </span>
</button>
`;

exports[`LimitPresetPriceButton renders 5 correctly, disabled true and selected false 1`] = `
.c0 {
  background-color: unset;
  color: #7D7D7D;
  padding: 2px 8px;
  height: 24px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background-color: transparent;
  border-radius: 999px;
  border: 1px solid #22222212;
}

<button
  class="c0"
>
  <span>
    +
    5%
  </span>
</button>
`;

exports[`LimitPresetPriceButton renders 10 correctly, disabled false and selected true 1`] = `
.c0 {
  background-color: unset;
  color: #222222;
  padding: 2px 8px;
  height: 24px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background-color: #22222212;
  border-radius: 999px;
  border: 1px solid #22222212;
  -webkit-text-decoration: none;
  text-decoration: none;
  cursor: pointer;
  -webkit-transition-duration: 125ms;
  transition-duration: 125ms;
}

.c0:hover {
  opacity: 0.6;
}

.c0:active {
  opacity: 0.4;
}

<button
  class="c0"
>
  <span>
    +
    10%
  </span>
</button>
`;
